WalletWasabi/WabiSabi/Client/CoinJoinClient.cs

```diff
@@ -91,6 +91,9 @@ public async Task<bool> StartRoundAsync(IEnumerable<SmartCoin> smartCoins, Round
 
 			var coinCandidates = SelectCoinsForRound(smartCoins, constructionState.Parameters);
 
+			var maxAvailableVsize = roundState.MaxVsizeAllocationPerAlice * coinCandidates.Count;
+			var amountDecomposerTask = Task.Run(() => new AmountDecomposer(coinCandidates, roundState, (int)maxAvailableVsize));
```

Comment:
Is this complexity only necessary for better performance? (Including the constructor of `AmountDecomposer` could be left stupid if you wouldn't want to gain performance?

Reply:
In a sense. But I would say algorithmic complexity, not performance.

Performance per se is out of scope for this PR, but it can also be improved, so I guess `Task.Run()` is unneeded, especially as the hard coded limit of 6 i put in CJC avoids triggering anything terrible. The main performance issue is memory consumption, which I know for a fact can be reduced very significantly, so that is not a long term concern but is not addressed in this PR.

The asymptotic behavior is more important because the underlying search space is exponential in size. Going from 3 to 4 is trivial. 4 to 5 can take a while, but using a static language means it's still very much doable (though python already struggles). Going up to 6 can easily consume several CPU minutes, especially for larger values.

Pruning reduces this asymptotic bound substantially, especially pruning by minimum value (but that can only be done at the last stage).

Finally, this API also supports efficient queries on ranges of values, which is important for decomposing many values (e.g. combinations of other users inputs), so the reason this work is done up front is because the results will be accessed repeatedly.